### PR TITLE
Fix edge case with optional dependencies

### LINKED ISSUE #572: Dependency resolution allows invalid tree in edge case with optional dependencies
If a plugin is registered with an optional dependency on TokenA, and later a plugin is registered with a required dependency on TokenA, and nothing gets registered on TokenA, fusion-core will not fail the resolution and instead will provide `undefined` for the injected value for TokenA. For example:

```js
function Main() {
  const app = new App();
  app.register(createPlugin({
    deps: {
      a: TokenA.Optional
    },
    provides({a}) => {
      console.log(a); // undefined
    }
  }));
  app.register(createPlugin({
    deps: {
      a: TokenA
    },
    provides({a}) => {
      // Dependency resolution should fail but it does not.
      console.log(a); // undefined
    }
  }));
  return app;
}
```

### PR DESCRIPTION
**Pull request by ganemone**
Fixes #572 and adds some skipped regression tests for #573

### CODE REVIEW

File: fusion-core/src/base-app.js [hunk 1]
```diff
@@ -1,5 +1,7 @@
 /** Copyright (c) 2018 Uber Technologies, Inc.
  *
+
```
Comment (AlexMSmithCA) [inline fusion-core/src/base-app.js]: Minor: remove new lines

File: fusion-core/src/__tests__/enhance.js [hunk 2]
```diff
@@ -166,6 +166,94 @@ tape('enhancement with a plugin with deps', t => {
   app.resolve();
 });
 
+// TODO(#573) This is a regression test which hasn't been fixed yet
```
Comment (AlexMSmithCA) [inline fusion-core/src/__tests__/enhance.js]: Thoughts on an additional plugin enhancer test where the enhancer's returned plugin has unregistered dependencies?  Something like:

```js
let app = new App('el', el => el);
type FnType = string => string;
const FnToken: Token<FnType> = createToken('FnType');
const EnhancerDepToken: Token<FnType> = createToken('EnhancerDepToken');
let BaseFnEnhancer = (fn: FnType): FusionPlugin<void, FnType> => {
  return createPlugin({
    deps: {
      addlFn: EnhancerDepToken,
    },
    provides: () => {
      return arg => {
        return fn(arg) + ' enhanced, and also ' + addlFn(arg);
      };
    },
  });
};
app.register(FnToken, msg => msg);
app.enhance(FnToken, BaseFnEnhancer);
t.throws(() => {
  app.resolve();
});
t.end();
```

(If we don't already cover this case)

Comment (ganemone) [reply fusion-core/src/__tests__/enhance.js]: We do have this test case

File: fusion-core/src/base-app.js [hunk 3]
```diff
@@ -172,73 +173,74 @@ class FusionApp {
...
-          this.register(token, undefined);
-        } else {
-          const dependents = Array.from(this.registered.entries());
+        // Early return if token is optional
+        const isOptional =
+          token instanceof TokenImpl && token.type === TokenType.Optional;
+        if (isOptional && (!enhancers || !enhancers.length)) {
+          return;
+        }
+        const dependents = Array.from(this.registered.entries());
```
Comment (AlexMSmithCA) [inline fusion-core/src/base-app.js]: Lines below here appear to be flagged for indentation as opposed to business logic changing.

**Comment by ganemone** [comment]
!merge

**Pull request merged**